File: services/ai-agents/app/domain/shared/aggregate_root.py
from __future__ import annotations

from .domain_event import DomainEvent


class AggregateRoot:
    """Records domain events. The use case pulls and publishes them after
    the transaction commits."""

    def __init__(self) -> None:
        self._events: list[DomainEvent] = []

    def _add_event(self, event: DomainEvent) -> None:
        self._events.append(event)

    def pull_events(self) -> list[DomainEvent]:
        out, self._events = self._events, []
        return out

File: services/auth/src/domain/shared/AggregateRoot.ts
import { DomainEvent } from "./DomainEvent";

// Base aggregate root. Records domain events that the application layer
// will dispatch after a successful transaction (write-then-publish).
export abstract class AggregateRoot {
  private _events: DomainEvent[] = [];

  protected addEvent(event: DomainEvent): void {
    this._events.push(event);
  }

  pullEvents(): DomainEvent[] {
    const out = this._events;
    this._events = [];
    return out;
  }
}
